extras/ws2812b/ws2812b.h

```diff
@@ -0,0 +1,210 @@
+ * end of data.
+ *
+ * @note
+ * The GPIO must be configured for output before trying
+ * to set colors!
+ *
+ * @attention
+ * Due to the precise timing required, interrupts are
+ * disabled until the data is sent.
+ *
+ * @author Ondřej Hruška, (c) 2016
+ *
+ * MIT License
+ */
+
+#ifndef WS2812B_DRV_H
+#define WS2812B_DRV_H
+
+#include <stdint.h>
+#include <stddef.h> // size_t
+
+#include "espressif/esp_common.h" // sdk_os_delay_us
+#include "esp/gpio.h"
+
+/**
+ * @brief Struct for easy manipulation of RGB colors.
+ *
+ * Set components in the xrgb.c.r (etc.) and you will get
+ * the hex in xrgb.num.
+ */
+typedef union {
+
+    /** Struct for ccess to individual color components */
+    struct __attribute__((packed)) {
+        uint8_t b;
+        uint8_t g;
+        uint8_t r;
+        uint8_t reserved; /**< padding field, could be used for brightness (?) */
+    } c;
+
+    /** RGB color as a single uint32_t */
+    uint32_t num;
+
+} xrgb_t;
+
+
+/**
+ * @brief Macro to compose a xrgb_t constant.
+ */
+#define rgb2xrgb(r, g, b) {.num = (((r) & 0xFF) << 16) | (((g) & 0xFF) << 8) | ((b) & 0xFF)}
+
+#define hex2xrgb(rrggbb) {.num = (rrggbb) & 0xFFFFFF}
+
+/**
+ * @brief Set a single RGB LED color, and display it.
+ *
+ * @param gpio_num : data line GPIO number
+ * @param rgb : RGB color - 0x00RRGGBB
+ */
+void ws2812b_set(uint8_t gpio_num, uint32_t rgb);
+
+
+/**
+ * @brief Set color of multiple RGB LEDs, and display it.
+ *
+ * @note
+ * Due to the precise timing required, interrupts are
+ * disabled until all data is sent (roughly 25*count + 50µs)
+ *
+ * @param gpio_num : data line GPIO number
+ * @param rgbs  : array of RGB colors in the 0x00RRGGBB format
+ * @param count : number of elements in the array
+ */
+void ws2812b_set_many(uint8_t gpio_num, uint32_t *rgbs, size_t count);
+
+
+/**
+ * @brief Turn a single WS2812B off
+ *
+ * This is a companion function to ws2812b_set().
+ *
+ * It sets the color to BLACK, thus effectively
+ * turning the RGB LED off.
+ *
+ * @param gpio_num : data line GPIO number
+ */
+void ws2812b_off(uint8_t gpio_num);
+
+
+
+///////////////////////////////////////////////////////////////////
+///
+/// The following part of the driver is used internally, and can
+/// also be used for "procedural generation" of colors.
+///
+/// - first call ws2812b_seq_start(),
+/// - then repeatedly ws2812b_seq_rgb() with your colors
+/// - and end the sequence with ws2812b_seq_end()
+///
+///////////////////////////////////////////////////////////////////
+
+
+
+// Ugly way to get short delays. Works for 80 MHz.
+
+// 400 ns
+#ifndef WS_SHORT_DELAY
+#define WS_SHORT_DELAY() for (volatile uint32_t __j = 1; __j > 0; __j--)
+#endif
+
+// 800 ns
+#ifndef WS_LONG_DELAY
+#define WS_LONG_DELAY()  for (volatile uint32_t __j = 4; __j > 0; __j--)
+#endif
+
+
+// inline attribute for functions that need zero overhead.
+#define INLINEFUNC static inline __attribute__((always_inline))
```

Comment:
Does this turn out to be necessary? I was a bit hooked on this kind of macro, a habit I picked up with old versions of avr-gcc that would fail to inline very obviously inlinable things!

But @foogod told me off and I went and pulled them all out of the codebase, without any major side effects that I can find. :)